build: validate options passed to configure

jbergstroem opened the pull request:
Validate some of the arguments passed to configurable options. Also, move defaults to optparse default.
I can't really vouch for options; just pulled them from comments. I've tested a few permutations, but not all.

jbergstroem commented:
https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/434/

jbergstroem commented:
Looks like there's some trickery going on at the arm architectures. Investigating.

mscdex commented:
+1

jbergstroem commented:
New build at https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/438/

Hopefully I didn't screw this one up..

rvagg commented:
love it, lgtm with squash

bnoordhuis commented on configure:
```diff
@@ -40,14 +48,16 @@ parser.add_option('--debug',
-         'Valid values are: arm, arm64, ia32, mips, mipsel, x32, x64')
+    choices=valid_arch,
+    help='CPU architecture to build for ({0})'.format(', '.join(
+        str(i) for i in valid_arch)))
```
You can shorten this to:

```python
'CPU architecture to build for ({0})'.format(', '.join(valid_arch))
```

Or if you don't care about python 3 compat:

```python
'CPU architecture to build for (%s)' % ', '.join(valid_arch)
```
```

jbergstroem replied:
Thanks. My Python is weak.

bnoordhuis commented on configure:
```diff
@@ -799,10 +809,6 @@ def configure_intl(o):
-  # sanity check localelist
-  if options.with_icu_locales and (with_intl != 'small-icu'):
-    print 'Error: --with-icu-locales only makes sense with --with-intl=small-icu'
-    sys.exit(1)
```
Can this check safely go?  Reading the source, I'm inclined to say it can't.

jbergstroem replied:
I think you're right. While looking at it the first time, I interpreted it as a boolean check (do we want icu?) which it obviously isn't.

I'm not super happy with the 100-or-so lines of icu in here; but I'll try to rewrite as little as possible to cater for my validation. Thanks for the catch.

jbergstroem commented:
Holding off until #1331 lands; no point in writing logic for shared v8 flags when we're removing it. Agree?

jbergstroem commented:
@bnoordhuis: I've looked closer at the icu branching. I'm still keen on removing the check altogether. The issue is that help text says that root,en is default - but it isn't set unless small-icu is in use. Removing the check will just make the build ignore `with_icu_locales` for anything else than small-icu. For small-icu we now set it through optparse. With your agreement I'll remove this logic.

Other than that, I've simplified the help text generation and rebased onto the v8 removal branch. PTAL.

jbergstroem commented:
@bnoordhuis this is what I have in mind:
```diff
diff --git configure configure
index 1d0e65a..148b208 100755
--- configure
+++ configure
@@ -787,13 +787,7 @@ def configure_intl(o):
     return
   # --with-intl=<with_intl>
   # set the default
-  if with_intl is None:
-    with_intl = 'none'  # The default mode of Intl
-  # sanity check localelist
-  if options.with_icu_locales and (with_intl != 'small-icu'):
-    print 'Error: --with-icu-locales only makes sense with --with-intl=small-icu'
-    sys.exit(1)
-  if with_intl == 'none' or with_intl is None:
+  if with_intl in (None, 'none'):
     o['variables']['v8_enable_i18n_support'] = 0
     return  # no Intl
   elif with_intl == 'small-icu':
@@ -801,8 +795,6 @@ def configure_intl(o):
     o['variables']['v8_enable_i18n_support'] = 1
     o['variables']['icu_small'] = b(True)
     with_icu_locales = options.with_icu_locales
-    if not with_icu_locales:
-      with_icu_locales = 'root,en'
     locs = set(with_icu_locales.split(','))
     locs.add('root')  # must have root
     o['variables']['icu_locales'] = string.join(locs,',')
```

bnoordhuis commented:
@jbergstroem Shouldn't with_icu_locales have a default in that case? Else the line that says `locs = set(with_icu_locales.split(','))` is going to error.

jbergstroem commented:
@bnoordhuis ah, sorry - should say `options.with_icu_locales`

bnoordhuis commented:
What I mean is, with_icu_locales or options.with_icu_locales is going to be None, so calling .split() on it won't work.

jbergstroem commented:
@bnoordhuis `with_icu_locales` will always be set to root,en unless you override it: https://github.com/iojs/io.js/pull/1335/files#diff-e2d5a00791bce9a01f99bc6fd613a39dR259

bnoordhuis commented:
Right you are.  Objection withdrawn.

jbergstroem commented:
@bnoordhuis good to merge?

bnoordhuis commented:
LGTM

jbergstroem commented:
Merged in fd90b33b94e6ce64fde193deca2e2aa6c774f558. Thanks for the feedback and helping me to get it in shape. Closing.
